stripe-ui-core/src/main/java/com/stripe/android/uicore/elements/TextFieldUI.kt

```diff
@@ -152,55 +157,58 @@ fun TextField(
-        if (fieldState == TextFieldStateConstants.Valid.Full && hasFocus) {
+        if (fieldState == TextFieldStateConstants.Valid.Full && hasFocus.value) {
             focusManager.moveFocus(nextFocusDirection)
         }
     }
 
     val autofillReporter = LocalAutofillEventReporter.current
 
+    var selection by remember {
+        mutableStateOf<TextRange?>(null)
+    }
+
     TextFieldUi(
-        value = value,
+        value = TextFieldValue(
+            text = value,
+            selection = selection ?: TextRange(value.length),
+        ),
         loading = loading,
         onValueChange = { newValue ->
-            val acceptInput = fieldState.canAcceptInput(value, newValue)
+            when (val newTextValue = newValue.text) {
+                value -> {
+                    selection = newValue.selection
+                }
+                else -> {
+                    val acceptInput = fieldState.canAcceptInput(value, newTextValue)
 
-            if (acceptInput) {
-                val newTextState = textFieldController.onValueChange(newValue)
+                    if (acceptInput) {
+                        selection = newValue.selection
 
-                if (newTextState != null) {
-                    onTextStateChanged(newTextState)
+                        val newTextState = textFieldController.onValueChange(newTextValue)
+
+                        if (newTextState != null) {
+                            onTextStateChanged(newTextState)
+                        }
+                    }
                 }
             }
         },
         onDropdownItemClicked = textFieldController::onDropdownItemClicked,
         modifier = modifier
-            .onPreviewKeyEvent { event ->
-                if (event.type == KeyEventType.KeyDown &&
-                    event.nativeKeyEvent.keyCode == KeyEvent.KEYCODE_DEL &&
-                    value.isEmpty()
-                ) {
-                    focusManager.moveFocus(previousFocusDirection)
-                    true
-                } else {
-                    false
-                }
-            }
-            .autofill(
-                types = listOfNotNull(textFieldController.autofillType),
-                onFill = {
-                    textFieldController.autofillType?.let { type ->
-                        autofillReporter(type.name)
-                    }
-                    textFieldController.onValueChange(it)
-                }
+            .onPreviewKeyEvent(
+                value = value,
+                focusManager = focusManager,
+                direction = previousFocusDirection
+            )
+            .onAutofill(
+                textFieldController = textFieldController,
+                autofillReporter = autofillReporter,
+            )
+            .onFocusChanged(
+                textFieldController = textFieldController,
+                hasFocus = hasFocus,
```

Comment:
`CyclomaticComplexMethod` detekt rule, opted to move the modifiers into their own functions.